app/src/androidTest/java/com/karumi/MainActivityTest.kt

```diff
@@ -0,0 +1,156 @@
+        val superHeroes = givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES, false)
+
+        startActivity()
+
+        RecyclerViewInteraction.onRecyclerView<SuperHero>(withId(R.id.recycler_view))
+                .withItems(superHeroes)
+                .check({ _, view, e ->
+                    matches(hasDescendant(allOf<View>(withId(R.id.iv_avengers_badge),
+                            withEffectiveVisibility(ViewMatchers.Visibility.GONE)))).check(view, e)
+                })
+    }
+
+    @Test
+    fun doesNotShowEmptyCaseIfThereAreSuperHeroes() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.tv_empty_case)).check(matches(not(isDisplayed())))
+    }
+
+    @Test
+    fun doesNotShowLoadingViewOnceSuperHeroesAreShown() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.progress_bar)).check(matches(not(isDisplayed())))
+    }
+
+    @Test
+    fun opensSuperHeroDetailActivityOnRecyclerViewItemTapped() {
+        val superHeroes = givenThereAreSomeSuperHeroes()
+        val superHeroIndex = 0
+        startActivity()
+
+        onView(withId(R.id.recycler_view))
+                .perform(actionOnItemAtPosition<RecyclerView.ViewHolder>(superHeroIndex, click()))
+
+        val superHeroSelected = superHeroes[superHeroIndex]
+        intended(hasComponent(SuperHeroDetailActivity::class.java.canonicalName))
+        intended(hasExtra("super_hero_name_key", superHeroSelected.name))
+    }
+
+    @Test
+    fun showsTheExactNumberOfSuperHeroes() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.recycler_view)).check(
+                matches(recyclerViewHasItemCount(ANY_NUMBER_OF_SUPER_HEROES)))
+    }
+
+    private fun givenThereAreSomeAvengers(numberOfAvengers: Int): List<SuperHero> =
+            givenThereAreSomeSuperHeroes(numberOfAvengers, true)
+
+    private fun givenThereAreSomeSuperHeroes(numberOfSuperHeroes: Int = ANY_NUMBER_OF_SUPER_HEROES, avengers: Boolean = false): List<SuperHero> {
+        val superHeroes = LinkedList<SuperHero>()
+        for (i in 0 until numberOfSuperHeroes) {
+            val superHeroName = "SuperHero - " + i
+            val superHeroPhoto = "https://i.annihil.us/u/prod/marvel/i/mg/c/60/55b6a28ef24fa.jpg"
+            val superHeroDescription = "Description Super Hero - " + i
+            val superHero = SuperHero(superHeroName, superHeroPhoto, avengers, superHeroDescription)
+            superHeroes.add(superHero)
+            on(repository.getByName(superHeroName)).thenReturn(superHero)
+        }
+        on(repository.getAllSuperHeroes()).thenReturn(superHeroes)
+        return superHeroes
+    }
+
+    private fun givenThereAreNoSuperHeroes() {
+        on(repository.getAllSuperHeroes()).thenReturn(emptyList<SuperHero>())
+    }
+
+    companion object {
```

Comment:
are we going to add companion objects in the top of the class or in the bottom?

Reply:
Let's keep it at the top of the classes just because 👍